File: ModoMercado/Core/Domain/Models/ListaDeCompras.swift
import Foundation

struct ListaDeCompras: Identifiable, Equatable, Codable {
    let id: UUID
    var nome: String
    var itens: [ItemDeLista]
    let criadoEm: Date
    var atualizadoEm: Date

    var totalEstimado: Decimal {
        itens.reduce(Decimal.zero) { $0 + $1.subtotalEstimado }
    }

    init(
        id: UUID = UUID(),
        nome: String,
        itens: [ItemDeLista] = [],
        criadoEm: Date = Date(),
        atualizadoEm: Date = Date()
    ) {
        self.id = id
        self.nome = nome
        self.itens = itens
        self.criadoEm = criadoEm
        self.atualizadoEm = atualizadoEm
    }
}

File: modo_mercado/lib/core/domain/models/lista_de_compras.dart
import 'package:equatable/equatable.dart';
import 'item_de_lista.dart';

class ListaDeCompras extends Equatable {
  final String id;
  final String nome;
  final List<ItemDeLista> itens;
  final DateTime criadoEm;
  final DateTime atualizadoEm;

  const ListaDeCompras({
    required this.id,
    required this.nome,
    required this.itens,
    required this.criadoEm,
    required this.atualizadoEm,
  });

  int get totalEstimadoCentavos =>
      itens.fold(0, (acc, item) => acc + item.subtotalEstimadoCentavos);

  ListaDeCompras copyWith({
    String? id,
    String? nome,
    List<ItemDeLista>? itens,
    DateTime? criadoEm,
    DateTime? atualizadoEm,
  }) {
    return ListaDeCompras(
      id: id ?? this.id,
      nome: nome ?? this.nome,
      itens: itens ?? this.itens,
      criadoEm: criadoEm ?? this.criadoEm,
      atualizadoEm: atualizadoEm ?? this.atualizadoEm,
    );
  }

  @override
  List<Object?> get props => [id, nome, itens, criadoEm, atualizadoEm];
}
